colcon_sanitizer_reports/sanitizer_log_parser.py

```diff
@@ -0,0 +1,195 @@
+        self.keys = _KeysFactory.make_keys(lines=lines, error_name=error_name)
+
+
+class _Section:
+    """Contents for a single Sanitizer output section."""
+
+    error_name: str
+    sub_sections: Tuple[_SubSection]
+
+    def __init__(self, *, lines: Tuple[str]) -> None:
+        # Section name comes after 'Sanitizer: ', and before any open paren or hex number.
+        self.error_name = re.match(
+            r'^.*Sanitizer: (?P<error_name>.+?)( \(| 0x[\dabcdef]+|\s*$)', lines[0]
+        ).groupdict()['error_name']
+
+        # Divide into _SubSections. SubSections begin with a line that is not indented.
+        sub_section_lines: List[str] = []
+        sub_sections: List[_SubSection] = []
+        for line in lines:
+            if re.match(r'^\S.*$', line) is not None and sub_section_lines:
+                sub_sections.append(
+                    _SubSection(lines=tuple(sub_section_lines), error_name=self.error_name)
+                )
+                sub_section_lines = []
+
+            sub_section_lines.append(line)
+        else:
+            sub_sections.append(
+                _SubSection(lines=tuple(sub_section_lines), error_name=self.error_name)
+            )
+
+        self.sub_sections = tuple(sub_sections)
+
+
+class SanitizerLogParser:
+    """Parses logged sanitizer errors and warnings and generates report."""
+
+    def __init__(self) -> None:
+        # Holds count of errors seen. This is what will be in the report.
+        self._counts = defaultdict(int)
+
+        # Current package and sections that are partially parsed
+        self._package: Optional[str] = None
+        self._section_lines_by_prefix: Dict[str: List[str]] = {}
+
+    @property
+    def csv(self) -> str:
+        """Return a csv representation of reported error/warnings."""
+        csv_f_out = StringIO()
+        writer = csv.writer(csv_f_out)
+        writer.writerow(['package', 'error_name', 'key', 'count'])
+        for (package, error_name, key), count in self._counts.items():
+            writer.writerow([package, error_name, key, count])
+
+        return csv_f_out.getvalue()
+
+    def set_package(self, package: Optional[str]) -> None:
+        self._package = package
+
+    def add_line(self, line: str) -> None:
+        """Generate report from log file lines."""
+        line = line.rstrip()
+
+        # If we have a new sanitizer section, start gathering lines for it.
+        match = re.match(r'^(?P<prefix>.*?)(==\d+==|)(WARNING|ERROR):.*Sanitizer:.*', line)
+        if match is not None:
+            prefix = match.groupdict()['prefix']
+            self._section_lines_by_prefix[prefix] = []
+
+        # Check if this line belongs to any of the sections we're currently building.
+        for prefix in self._section_lines_by_prefix.keys():
+            match = re.match(r'^{prefix}(?P<line>.*)$'.format(prefix=re.escape(prefix)), line)
+            if match is not None:
```

Comment:
`if not match` ?

Reply:
In this case it would be fine since an re.Pattern always evaluates to True. I'd prefer to follow the pep8 recommendation (not requirement) to always test Optional types for None-or-something instead of True or False. Kinda feels like I'm asking for a subtle bug somewhere in my life otherwise.

Reply:
sgtm